File: examples/backend/java/UserService.java
package com.example.user;

import java.util.Optional;

/**
 * Pattern points:
 * - Java records for DTOs; entities never leave the service boundary
 * - Constructor injection (no field @Autowired)
 * - Constructor takes the interface — Mockito swaps it in tests
 */
public class UserService {

    public record UserDTO(long id, String username, String email) {}

    public static class UserNotFoundException extends RuntimeException {
        public UserNotFoundException(long id) {
            super("user %d not found".formatted(id));
        }
    }

    public interface UserRepository {
        Optional<UserDTO> findById(long id);
        Optional<UserDTO> findByUsername(String username);
        UserDTO insert(String username, String email);
    }

    private final UserRepository repository;

    public UserService(UserRepository repository) {
        this.repository = repository;
    }

    public UserDTO getUser(long id) {
        return repository.findById(id)
                .orElseThrow(() -> new UserNotFoundException(id));
    }

    public UserDTO register(String username, String email) {
        if (username == null || username.isBlank() || username.length() > 64) {
            throw new IllegalArgumentException("username must be 1-64 characters");
        }
        if (email == null || !email.contains("@")) {
            throw new IllegalArgumentException("invalid email address");
        }
        repository.findByUsername(username).ifPresent(existing -> {
            throw new IllegalArgumentException("username '%s' already taken".formatted(username));
        });
        return repository.insert(username, email);
    }
}

File: examples/backend/php/UserService.php
<?php

declare(strict_types=1);

namespace App\User;

use InvalidArgumentException;

/**
 * Pattern points:
 * - strict_types on every file
 * - readonly promoted constructor properties, constructor injection
 * - DTO record-style class at the boundary
 * - #[\Override] on interface implementations
 */
final class UserDTO
{
    public function __construct(
        public readonly int $id,
        public readonly string $username,
        public readonly string $email,
    ) {}
}

interface UserRepository
{
    public function findById(int $id): ?UserDTO;

    public function findByUsername(string $username): ?UserDTO;

    public function insert(string $username, string $email): UserDTO;
}

final class UserNotFoundException extends \RuntimeException
{
    public function __construct(int $id)
    {
        parent::__construct(sprintf('user %d not found', $id));
    }
}

final class UserService
{
    public function __construct(
        private readonly UserRepository $repository,
    ) {}

    public function getUser(int $id): UserDTO
    {
        return $this->repository->findById($id)
            ?? throw new UserNotFoundException($id);
    }

    public function register(string $username, string $email): UserDTO
    {
        if ($username === '' || strlen($username) > 64) {
            throw new InvalidArgumentException('username must be 1-64 characters');
        }
        if (!str_contains($email, '@')) {
            throw new InvalidArgumentException('invalid email address');
        }
        if ($this->repository->findByUsername($username) !== null) {
            throw new InvalidArgumentException("username '$username' already taken");
        }

        return $this->repository->insert($username, $email);
    }
}
